design/keycloak.x/storage-architecture.md

```diff
@@ -0,0 +1,129 @@
+to have e.g. realm definitions stored in files and cached in Infinispan,
+clients stored in etcd store and cached in plain hashmaps and sessions
+shared in Infinispan caches. (This is indeed overstretched but you get
+the picture.) With this split it comes that the references to other
+objects would be loose, and it will be Keycloak responsibility to keep
+referential integrity.
+
+### First-class objects
+
+First-class objects are objects that have a common store settings, and
+their separation is logical  from the perspective of objects handled in
+Keycloak. The first-class objects are depicted below in the color boxes.
+A **realm** contains all data related to the realm that almost never
+changes (realm name and other attributes, identity providers, user
+federation, flows setup of keys etc.). A **client** contains similarly
+all the data related to the client, and **client scope** description of
+the client scope attachable to a client. **Authorization** store contains
+definitions of resources, scopes, policies and permissions. **User**,
+**role**, and **group**, as well as **event** are self-explanatory.
+Sessions have data stored separately for **authentication sessions** and
+**user sessions**.
+
+![First-class objects](img/storage-first-class-objects.png)
+
+### Store configuration
+
+For each object type, a separate store can be configured. The store
+allows configuring persistence and caching independently. For the current
+store, JPA would be used for persistence and Infinispan for caching in
+case of realms, roles etc., and Infinispan would be used for persistence
+and passthrough (i.e. nothing) for caching in case of sessions.
+
+By separating the configuration of the two, it is possible to create
+often-used templates, e.g. JPA+Infinispan, Files/YAML+Infinispan,
+etcd/JSON+passthrough etc.
+
+Caching requires some kind of event support from the underlying store to
+be able to invalidate the cached objects.
+
+![Layers in store.x](img/storage-caching-layers.svg)
+
+## Physical layout
+
+Since the references between the objects will be loose, the benefits of a
+relational database diminish, and most of the work with querying and
+maintaining indexes (if any) will be done in Java. The objects in
+Keycloak JVM have three representations:
+
+1. Logical representation (a.k.a. `*Model` and `*Adapter` in current
+   codebase) - the representation contains reference to KeycloakSession and
+   thus depend on the state of the objects (realms, sessions etc.)
+   throughout the period of handling the request
+
+2. Stateless representation (similar to aggregated `*Entity`) - the
+   representation contains IDs rather than ability to provide logical
+   representation of the object. The crucial feature of each field type is
+   that its representation is serializable, and is independent of the
+   datastore format.
+
+3. Storage representation - datastore-dependent representation, e.g.
```

Comment:
What will the be "future" workflow if I wanted to add a new field to the Realm? What objects will I need to modify?

From my understanding, I would need to add it to the logical representation and all storage representations as well (Infinispan, JPA, YAML, and others). How do we plan to keep all this places consistent?

Reply:
This is a very good question for an actual store. You should need to modify the model and a common ancestor, that should be it.
For storage, it will depend on the level of the particular storage implementation, but every storage will have support for storing unknown fields in some form, e.g. in JSON, and also all the fields will need to be serializable to a String for the same reason. If desirable, the storage can adjust physical layout to provide the best fit for a particular field, e.g. by storing its value in a column in a table in case of JPA.